### DISCUSSION TITLE
What could be the reason that BeforeAllCallback runs before Each Test

### DISCUSSION THREAD

**Discussion by pratyushdeshpande** [junit-team/junit-framework#3158] [General]
I am running Parallel test with Selenium and I have implemented BeforeEachCallback, AfterEachCallback, BeforeAllCallback and also ExtensionContext.Store.CloseableResource and used this class as @ExtendsWith annotation with all my test Classes. 
I have implemented a Different Namespace, e.g. Namespace.create("QATEST") and putting a resources in the namespace.

My All tests are Parameterized tests with custom annotation created using @ArgumentsSource . 

When I execute my tests, even if sequentially or parallelly, I see that BeforeAll block of code is executed before every Parameterized test execution similar to how beforeEach executes. Also same behaviour observed for Closeable Resource.

What could be the reason for this behaviour/am I missing something? and what could be done to change this to normal behaviour of executing BeforeAll block only once before executing all tests and closeableResouce block to execute once after all tests are executed. 

Using:
Junit5 : 5.8.2
Selenium 4: 4.7.2

**Comment by sbrannen** [comment]
A `BeforeAllCallback` will be invoked once per test class when running all test classes within the same test suite (i.e., same JVM process).

Are you perhaps configuring your build to **fork** test execution per test class?

**Reply by pratyushdeshpande** [reply]
Yes I am forking per test class, but BeforeAllCallback is executed before each Test method and not before each Test class, thqt is why I am confused If I missed something here.

**Reply by sbrannen** [reply]
Are you perhaps forking for each test (method) instead of each test class?

If not, we'd need to see a minimal example project that demonstrates the issue.

Though, such questions are typically better asked on Stack Overflow since it does not appear to be an issue with JUnit itself.
